### PR TITLE
TLS 1.3 code integrated into ssl_client2 /ssl_server2 examples

### PR DESCRIPTION
**Pull request by hannestschofenig**
Removed tls13_client.c and tls13_server.c examples.

### CODE REVIEW

File: programs/ssl/ssl_client2.c [hunk 1]
```diff
@@ -547,6 +597,14 @@ struct options
                                  * after renegotiation                      */
     int reproducible;           /* make communication reproducible          */
     int skip_close_notify;      /* skip sending the close_notify alert      */
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL) 
+	const char *named_groups_string;    /* list of named groups             */
```
Comment (gbryant-arm) [inline programs/ssl/ssl_client2.c]: Coding style: indentation

File: programs/ssl/ssl_client2.c [hunk 2]
```diff
@@ -1194,7 +1274,22 @@ int main( int argc, char *argv[] )
     mbedtls_net_init( &server_fd );
     mbedtls_ssl_init( &ssl );
     mbedtls_ssl_config_init( &conf );
-    memset( &saved_session, 0, sizeof( mbedtls_ssl_session ) );
+
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL) 
+
+#if defined(MBEDTLS_SSL_NEW_SESSION_TICKET)
+	mbedtls_ssl_init_client_ticket(&ticket);
```
Comment (gbryant-arm) [inline programs/ssl/ssl_client2.c]: Coding style: space between opening and closing parentheses

File: programs/ssl/ssl_client2.c [hunk 3]
```diff
@@ -1194,7 +1274,22 @@ int main( int argc, char *argv[] )
...
+#endif /* MBEDTLS_SSL_NEW_SESSION_TICKET */
+
+#if defined(MBEDTLS_ECP_C)
+	memset((void *)named_groups_list, MBEDTLS_ECP_DP_NONE, sizeof(named_groups_list));
```
Comment (gbryant-arm) [inline programs/ssl/ssl_client2.c]: Coding style: space between opening and closing parentheses

File: programs/ssl/ssl_client2.c [hunk 4]
```diff
@@ -1534,6 +1639,41 @@ int main( int argc, char *argv[] )
...
+#if defined(MBEDTLS_ECP_C)
+		else if( strcmp( p, "named_groups" ) == 0 )
+			opt.named_groups_string = q;
+		else if( strcmp( p, "key_share_named_groups" ) == 0 )
+			opt.key_share_named_groups_string = q;
+#endif /* MBEDTLS_ECP_C */
+
+        else if( strcmp( p, "key_exchange_modes" ) == 0 ) 
+        {
+			if( strcmp( q, "psk" ) == 0 )
```
Comment (gbryant-arm) [inline programs/ssl/ssl_client2.c]: Coding style: indentation

File: programs/ssl/ssl_client2.c [hunk 5]
```diff
@@ -1616,6 +1773,14 @@ int main( int argc, char *argv[] )
                 opt.max_version = MBEDTLS_SSL_MINOR_VERSION_3;
                 opt.transport = MBEDTLS_SSL_TRANSPORT_DATAGRAM;
             }
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL) 
+			else if( strcmp( q, "dtls1_3" ) == 0 )
+			{
+				opt.min_version = MBEDTLS_SSL_MINOR_VERSION_4;
+				opt.max_version = MBEDTLS_SSL_MINOR_VERSION_4;
+                opt.transport = MBEDTLS_SSL_TRANSPORT_DATAGRAM;
```
Comment (gbryant-arm) [inline programs/ssl/ssl_client2.c:1775]: Coding style: indentation

File: programs/ssl/ssl_client2.c [hunk 6]
```diff
@@ -1961,6 +2126,84 @@ int main( int argc, char *argv[] )
...
+
+		/* Leave room for a final NULL in named_groups_list */
+		while( i < (int) sizeof named_groups_list - 1 && *p != '\0' )
+		{
+
+			while( *p != ',' && *p != '\0' )
+				p++;
+
+			if( *p == ',' || *p == '\0' ) 
+            {
```
Comment (gbryant-arm) [inline programs/ssl/ssl_client2.c:2176]: Coding style: indentation

File: programs/ssl/ssl_client2.c [hunk 7]
```diff
@@ -2178,17 +2421,25 @@ int main( int argc, char *argv[] )
...
     }
+#endif /* !MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL */        
 
     if( opt.context_crt_cb == 0 )
         mbedtls_ssl_conf_verify( &conf, my_verify, NULL );
 
     memset( peer_crt_info, 0, sizeof( peer_crt_info ) );
+    
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL) && defined(MBEDTLS_ZERO_RTT)
+	mbedtls_ssl_conf_early_data(&conf, opt.early_data, early_data, strlen(early_data), NULL);
```
Comment (gbryant-arm) [inline programs/ssl/ssl_client2.c]: Coding style: space between opening and closing parentheses

File: programs/ssl/ssl_client2.c [hunk 8]
```diff
@@ -2275,6 +2526,11 @@ int main( int argc, char *argv[] )
                                   : MBEDTLS_SSL_CBC_RECORD_SPLITTING_DISABLED );
 #endif
 
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL) 
+    // Configure key exchange mode
+    mbedtls_ssl_conf_ke(&conf, opt.key_exchange_modes);
```
Comment (gbryant-arm) [inline programs/ssl/ssl_client2.c]: Coding style: space between opening and closing parentheses

File: programs/ssl/ssl_client2.c [hunk 9]
```diff
@@ -2347,12 +2603,19 @@ int main( int argc, char *argv[] )
 #endif
 
 #if defined(MBEDTLS_ECP_C)
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL)
+    if (named_groups_list[0]!= MBEDTLS_ECP_DP_NONE)
+		mbedtls_ssl_conf_curves(&conf, named_groups_list);
```
Comment (gbryant-arm) [inline programs/ssl/ssl_client2.c]: Coding style: space between opening and closing parentheses

File: programs/ssl/ssl_client2.c [hunk 10]
```diff
@@ -2347,12 +2603,19 @@ int main( int argc, char *argv[] )
...
+	if (key_share_named_groups_list[0] != MBEDTLS_ECP_DP_NONE)
+		mbedtls_ssl_conf_key_share_curves(&conf, key_share_named_groups_list);
```
Comment (gbryant-arm) [inline programs/ssl/ssl_client2.c]: Coding style: space between opening and closing parentheses

File: programs/ssl/ssl_client2.c [hunk 11]
```diff
@@ -2936,6 +3236,39 @@ int main( int argc, char *argv[] )
...
+		ret = mbedtls_ssl_get_client_ticket( &ssl, &ticket );
+
+		if( ret < 0 )
+		{
+			mbedtls_printf( " failed\n  ! mbedtls_ssl_get_ticket returned -0x%x\n\n",
+                                       (unsigned) -ret );
+			goto exit;
+		}
+		else if( ret == 1 ) 
+        {
```
Comment (gbryant-arm) [inline programs/ssl/ssl_client2.c:3255]: Coding style: indentation

File: programs/ssl/ssl_client2.c [hunk 12]
```diff
@@ -2936,6 +3236,39 @@ int main( int argc, char *argv[] )
...
+			// no ticket available - we cannot re-connect
+			opt.reconnect = 0;
+			mbedtls_printf( "no ticket available\n" );
+
+		}
+		else if( ret == 0 ) 
+        {
```
Comment (gbryant-arm) [inline programs/ssl/ssl_client2.c:3262]: Coding style: indentation

File: programs/ssl/ssl_client2.c [hunk 13]
```diff
@@ -3250,6 +3583,21 @@ int main( int argc, char *argv[] )
...
+		 * At a minimum we need to configure the psk (with the resumption_secret)
+		 * and the psk_identity (with the ticket).
+		 */
+		if(( ret = mbedtls_ssl_conf_client_ticket( &ssl, &ticket )) != 0 )
+		{
+			mbedtls_printf(" failed\n  ! mbedtls_ssl_conf_client_ticket returned %d\n\n", ret);
+			goto exit;
+		}
+        // enable resumption
+		mbedtls_ssl_conf_client_ticket_enable(&ssl);
```
Comment (gbryant-arm) [inline programs/ssl/ssl_client2.c]: Coding style: space between opening and closing parentheses

File: programs/ssl/ssl_client2.c [hunk 14]
```diff
@@ -3330,7 +3679,13 @@ int main( int argc, char *argv[] )
     psa_destroy_key( key_slot );
 #endif
 #endif
+
+#if defined(MBEDTLS_SSL_NEW_SESSION_TICKET)
+    mbedtls_ssl_del_client_ticket(&ticket);
```
Comment (gbryant-arm) [inline programs/ssl/ssl_client2.c]: Coding style: space between opening and closing parentheses

File: programs/ssl/ssl_server2.c [hunk 15]
```diff
@@ -616,6 +678,9 @@ struct options
     int trunc_hmac;             /* accept truncated hmac?                   */
     int tickets;                /* enable / disable session tickets         */
     int ticket_timeout;         /* session ticket lifetime                  */
+#if defined(MBEDTLS_SSL_NEW_SESSION_TICKET)
+	mbedtls_ssl_ticket_flags ticket_flags;   /* ticket flags                */
```
Comment (gbryant-arm) [inline programs/ssl/ssl_server2.c]: Coding style: indentation

File: programs/ssl/ssl_server2.c [hunk 16]
```diff
@@ -1328,6 +1398,24 @@ int psk_callback( void *p_info, mbedtls_ssl_context *ssl,
 }
 #endif /* MBEDTLS_KEY_EXCHANGE_SOME_PSK_ENABLED */
 
+#if defined(MBEDTLS_ZERO_RTT)
+/*
+* Early data callback.
+*/
+int early_data_callback(mbedtls_ssl_context *ssl,
+	unsigned char *buffer, size_t len)
```
Comment (gbryant-arm) [inline programs/ssl/ssl_server2.c]: Coding style: indentation

File: programs/ssl/ssl_server2.c [hunk 17]
```diff
@@ -1328,6 +1398,24 @@ int psk_callback( void *p_info, mbedtls_ssl_context *ssl,
...
+{
+	// In this example we don't need access to the SSL structure
+       ((void) ssl);
+
+	if (len > 0 && buffer != NULL) {
+		buffer[len] = '\0';
+		mbedtls_printf(" %zu bytes early data received: %s\n", len, (char *)buffer);
```
Comment (gbryant-arm) [inline programs/ssl/ssl_server2.c]: Coding style: space between opening and closing parentheses

File: programs/ssl/ssl_server2.c [hunk 18]
```diff
@@ -2337,8 +2523,13 @@ int main( int argc, char *argv[] )
         else if( strcmp( p, "cookies" ) == 0 )
         {
             opt.cookies = atoi( q );
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL) 
+            if( opt.cookies < -1 || opt.cookies > 2)
```
Comment (gbryant-arm) [inline programs/ssl/ssl_server2.c]: Coding style: space between opening and closing parentheses

File: programs/ssl/ssl_server2.c [hunk 19]
```diff
@@ -3127,8 +3394,42 @@ int main( int argc, char *argv[] )
...
+                                        mbedtls_ctr_drbg_random, &ctr_drbg ) ) != 0 )
+        {
+            mbedtls_printf( " failed\n  ! mbedtls_ssl_cookie_setup returned %d\n\n", ret );
+            goto exit;
+        }
+
+        if( opt.cookies == 1 ) 
+        {
+            mbedtls_ssl_conf_cookies( &conf, mbedtls_ssl_cookie_write, mbedtls_ssl_cookie_check,
+            &cookie_ctx, MBEDTLS_SSL_FORCE_RR_CHECK_OFF );
```
Comment (gbryant-arm) [inline programs/ssl/ssl_server2.c]: Coding style: indentation

File: programs/ssl/ssl_server2.c [hunk 20]
```diff
@@ -3127,8 +3394,42 @@ int main( int argc, char *argv[] )
...
+        }
+
+        if( opt.cookies == 2 ) 
+        {
+            mbedtls_ssl_conf_cookies( &conf, mbedtls_ssl_cookie_write, mbedtls_ssl_cookie_check,
+            &cookie_ctx, MBEDTLS_SSL_FORCE_RR_CHECK_ON );
```
Comment (gbryant-arm) [inline programs/ssl/ssl_server2.c]: Coding style: indentation

File: programs/ssl/ssl_server2.c [hunk 21]
```diff
@@ -3327,12 +3629,17 @@ int main( int argc, char *argv[] )
 #endif
 
 #if defined(MBEDTLS_ECP_C)
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL)
+	if (named_groups_list[0] != MBEDTLS_ECP_DP_NONE)
+		mbedtls_ssl_conf_curves(&conf, named_groups_list);
```
Comment (gbryant-arm) [inline programs/ssl/ssl_server2.c]: Coding style: space between opening and closing parentheses

**Review by gbryant-arm — COMMENTED** [review]
Coding style review essentially...

File: programs/ssl/ssl_client2.c [hunk 22]
```diff
@@ -389,6 +397,30 @@ int main( void )
...
+#if defined(MBEDTLS_ZERO_RTT)
+#define USAGE_EARLY_DATA \
+    "    early_data=%%s        default: (library default: disabled)\n"      \
+    "                        options: disabled, enabled\n"
+#else
+#define USAGE_EARLY_DATA ""
+#endif /* MBEDTLS_ZERO_RTT */
+
+#if defined(MBEDTLS_ECP_C)
+#define USAGE_NAMED_GROUP \
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c:420]: Minor: This should likely be guarded by `MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL`, too.

File: programs/ssl/ssl_client2.c [hunk 23]
```diff
@@ -389,6 +397,30 @@ int main( void )
...
+    "    named_groups=%%s    default: secp256r1, secp384r1\n"      \
+    "                        options: secp256r1, secp384r1, secp521r1, all\n"
+#else
+#define USAGE_NAMED_GROUP ""
+#endif /* MBEDTLS_ECP_C */
+
+#if defined(MBEDTLS_ECP_C)
+#define USAGE_KEY_SHARE_NAMED_GROUPS \
+    "    key_share_named_groups=%%s    default: secp256r1, secp384r1\n"      \
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c:429]: Minor: This should likely be guarded by MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL, too.

File: programs/ssl/ssl_client2.c [hunk 24]
```diff
@@ -389,6 +397,30 @@ int main( void )
...
+    "                        options: secp256r1, secp384r1, secp521r1, all\n"
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c:430]: Indicate in the help string that this is TLS 1.3 only?

File: programs/ssl/ssl_client2.c [hunk 25]
```diff
@@ -389,6 +397,30 @@ int main( void )
...
+#define USAGE_EARLY_DATA \
+    "    early_data=%%s        default: (library default: disabled)\n"      \
+    "                        options: disabled, enabled\n"
+#else
+#define USAGE_EARLY_DATA ""
+#endif /* MBEDTLS_ZERO_RTT */
+
+#if defined(MBEDTLS_ECP_C)
+#define USAGE_NAMED_GROUP \
+    "    named_groups=%%s    default: secp256r1, secp384r1\n"      \
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c:421]: Indicate in the help string that this is TLS 1.3 only?

File: programs/ssl/ssl_client2.c [hunk 26]
```diff
@@ -389,6 +397,30 @@ int main( void )
 #define USAGE_SERIALIZATION ""
 #endif
 
+#if defined(MBEDTLS_ZERO_RTT)
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c]: Same here: TLS 1.3 guard + info string indicating that this is TLS 1.3 specific?

File: programs/ssl/ssl_client2.c [hunk 27]
```diff
@@ -216,10 +215,19 @@ int main( void )
 #endif /* MBEDTLS_SSL_DTLS_CONNECTION_ID */
 
 #if defined(MBEDTLS_KEY_EXCHANGE_SOME_PSK_ENABLED)
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL) 
+#define USAGE_PSK_RAW                                               \
+    "    psk=%%s                  default: \"\" (disabled)\n"     \
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c:221]: Minor: The old and new branches share the documentation for `psk` and `psk_identity`. Remove this duplication?

File: programs/ssl/ssl_client2.c [hunk 28]
```diff
@@ -466,7 +501,22 @@ int main( void )
     "                                is printed if it is defined\n"     \
     USAGE_SERIALIZATION                                     \
     " acceptable ciphersuite names:\n"
-
+#define USAGE5 \
+    "    arc4=%%d             default: (library default: 0)\n" \
+    "    allow_sha1=%%d       default: 0\n"                             \
+    "    min_version=%%s      default: (library default: tls1)\n"       \
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c]: Where does this block come from? It seems unrelated to this PR.

File: programs/ssl/ssl_client2.c [hunk 29]
```diff
@@ -960,6 +1018,7 @@ static int my_verify( void *data, mbedtls_x509_crt *crt,
     return( 0 );
 }
 
+#if !defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL)
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c]: Why is this guard being added? Ultimately, we want the example application to be able to serve both TLS 1.2 and TLS 1.3, and i nthis case, `ssl_sig_hashes_for_test[]` would be needed even if `MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL` is enabled.

I understand that there the prototype does currently not allow for simultaneous enabling of TLS 1.2 and TLS 1.3, but I think we should try to not amplify this gap further, since it's one of the main obstacles towards upstreaming.

File: programs/ssl/ssl_client2.c [hunk 30]
```diff
@@ -1135,7 +1195,7 @@ int main( int argc, char *argv[] )
     unsigned char alloc_buf[MEMORY_HEAP_SIZE];
 #endif
 
-#if defined(MBEDTLS_ECP_C)
+#if defined(MBEDTLS_ECP_C) && !defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL)
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c:1193]: Same here.

File: programs/ssl/ssl_client2.c [hunk 31]
```diff
@@ -1149,14 +1209,34 @@ int main( int argc, char *argv[] )
     psa_status_t status;
 #endif
 
-#if defined(MBEDTLS_X509_CRT_PARSE_C)
+#if defined(MBEDTLS_X509_CRT_PARSE_C) && !defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL)
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c]: Same here.

File: programs/ssl/ssl_client2.c [hunk 32]
```diff
@@ -1225,7 +1320,11 @@ int main( int argc, char *argv[] )
         mbedtls_printf( USAGE1 );
         mbedtls_printf( USAGE2 );
         mbedtls_printf( USAGE3 );
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL)
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c]: It's easier to read if the relevant configuration dependent logic is within the definition of `USAGE_X`, so we can simply print all `USAGE_X` strings here.

File: programs/ssl/ssl_client2.c [hunk 33]
```diff
@@ -1262,6 +1361,12 @@ int main( int argc, char *argv[] )
     opt.key_file            = DFL_KEY_FILE;
     opt.key_opaque          = DFL_KEY_OPAQUE;
     opt.psk                 = DFL_PSK;
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL) 
+	opt.key_exchange_modes = DFL_KEY_EXCHANGE_MODES;
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c]: Minor: Indentation

File: programs/ssl/ssl_client2.c [hunk 34]
```diff
@@ -1494,7 +1599,7 @@ int main( int argc, char *argv[] )
         else if( strcmp( p, "tickets" ) == 0 )
         {
             opt.tickets = atoi( q );
-            if( opt.tickets < 0 || opt.tickets > 2 )
+            if( opt.tickets < 0 || opt.tickets > 1 )
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c:1596]: Why was this change made?

File: programs/ssl/ssl_client2.c [hunk 35]
```diff
@@ -1534,6 +1639,41 @@ int main( int argc, char *argv[] )
                 default: goto usage;
             }
         }
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL) 
+
+#if defined(MBEDTLS_ZERO_RTT)
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c:1638]: Style: Tabs

File: programs/ssl/ssl_client2.c [hunk 36]
```diff
@@ -1546,6 +1686,11 @@ int main( int argc, char *argv[] )
             else if( strcmp( q, "tls1_2" ) == 0 ||
                      strcmp( q, "dtls1_2" ) == 0 )
                 opt.min_version = MBEDTLS_SSL_MINOR_VERSION_3;
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL) 
+            else if( strcmp( q, "tls1_3" ) == 0 ||
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c:1684]: Style: Tabs

File: programs/ssl/ssl_client2.c [hunk 37]
```diff
@@ -1961,6 +2126,84 @@ int main( int argc, char *argv[] )
...
 
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL) && defined(MBEDTLS_ECP_C)
+	if( opt.named_groups_string != NULL )
+	{
+		p = (char *)opt.named_groups_string;
+		i = 0;
+		start = p;
+
+		/* Leave room for a final NULL in named_groups_list */
+		while( i < (int) sizeof named_groups_list - 1 && *p != '\0' )
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c]: Style: Brackets after `sizeof`

File: programs/ssl/ssl_client2.c [hunk 38]
```diff
@@ -1961,6 +2126,84 @@ int main( int argc, char *argv[] )
     }
 #endif /* MBEDTLS_SSL_ALPN */
 
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL) && defined(MBEDTLS_ECP_C)
+	if( opt.named_groups_string != NULL )
+	{
+		p = (char *)opt.named_groups_string;
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c]: Style: `(char*) xxx` instead of `(char *)x`

File: programs/ssl/ssl_client2.c [hunk 39]
```diff
@@ -1961,6 +2126,84 @@ int main( int argc, char *argv[] )
...
+				}
+				else goto usage;
+				start = p;
+			}
+		}
+
+		if( i == 0 ) goto usage;
+	}
+
+	if (opt.key_share_named_groups_string != NULL)
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c]: Style: Spacing

File: programs/ssl/ssl_client2.c [hunk 40]
```diff
@@ -1961,6 +2126,84 @@ int main( int argc, char *argv[] )
...
+	{
+		p = (char *)opt.key_share_named_groups_string;
+		i = 0;
+		start = p;
+
+		/* Leave room for a final NULL in named_groups_list */
+		while (i < (int) sizeof key_share_named_groups_list - 1 && *p != '\0')
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c]: Style: Parentheses

File: programs/ssl/ssl_client2.c [hunk 41]
```diff
@@ -1961,6 +2126,84 @@ int main( int argc, char *argv[] )
...
+				start = p;
+			}
+		}
+
+		if( i == 0 ) goto usage;
+	}
+
+	if (opt.key_share_named_groups_string != NULL)
+	{
+		p = (char *)opt.key_share_named_groups_string;
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c]: Style: As above.

File: programs/ssl/ssl_client2.c [hunk 42]
```diff
@@ -2178,17 +2421,25 @@ int main( int argc, char *argv[] )
 #if defined(MBEDTLS_X509_CRT_PARSE_C)
     /* The default algorithms profile disables SHA-1, but our tests still
        rely on it heavily. */
+
+#if !defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL)
```
Comment (hanno-becker) [inline programs/ssl/ssl_client2.c]: As above: We shouldn't further exacerbate the incompatibility between TLS 1.2 and TLS 1.3 builds. This code path should be enabled even if `MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL` is set.

**Review by hanno-becker — CHANGES_REQUESTED** [review]
Thanks @hannestschofenig for working on consolidating the tests and example programs again, that's an important step to being able to upstream further parts of the prototype.

There are numerous style issues that need to be addressed. Apart from that, there are a few occasions where TLS 1.2 specific code in the example programs is guarded by `MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL`. As mentioned in the comment, while I understand that currently the 1.3 prototype is incompatible with 1.2, I think we should make an effort not to widen that incompatibility further - concretely, I don't think we should remove any functionality from the existing example program if `MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL` is set.

File: programs/ssl/ssl_client2.c (hunk 27 above)
Comment (hannestschofenig) [reply programs/ssl/ssl_client2.c:221]: In 1.3 we also have to define the key_exchange_modes along with it.

Comment (hanno-becker) [reply programs/ssl/ssl_client2.c:221]: Yep, but the duplication of `psk` and `psk_identity` can be avoided, or not?

File: programs/ssl/ssl_client2.c (hunk 34 above)
Comment (hannestschofenig) [reply programs/ssl/ssl_client2.c:1596]: The values defined for tickets are 0 or 1.
